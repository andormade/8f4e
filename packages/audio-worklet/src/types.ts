export interface MidiModuleAddresses {
	moduleId: string;
	noteAddress: number | undefined;
	channelAddress: number | undefined;
	noteOnOffAddress: number | undefined;
	velocityAddress: number | undefined;
}

export interface MidiCCModuleAddresses {
	moduleId: string;
	valueAddress: number | undefined;
	channelAddress: number | undefined;
	selectedCCAddress: number | undefined;
}

export interface RNBOModule {
	patcherId: string;
	paramAdresses: number[];
}
