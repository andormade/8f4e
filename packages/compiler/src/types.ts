export interface CompiledModule {
	functionBody: number[];
	initialMemory: number[];
	memoryAddresses: { address: number; id: string; default?: number }[];
	moduleId: string;
	offset: number;
}

export type MemoryBuffer = Int32Array;

export interface Connection {
	fromModuleId: string;
	fromConnectorId: string;
	toModuleId: string;
	toConnectorId: string;
}

export type RelativeAddressCalculator = (nthWord: number) => number;

export type ModuleGenerator<TConfig = unknown> = (
	moduleId: string,
	offset: RelativeAddressCalculator,
	initialConfig?: TConfig
) => CompiledModule;

export type EngineConfig = Record<string, number | string | EngineConfig[]>;

export interface Engine {
	name: string;
	config: EngineConfig;
}

export type ModuleState = Record<string, any>;

export type ModuleStateExtractor<T> = (memoryBuffer: MemoryBuffer, moduleAddress: number) => T;
export type ModuleStateInserter<T> = (moduleState: T, memoryBuffer: MemoryBuffer, moduleAddress: number) => void;

export interface Module {
	id: string;
	engine: Engine;
	state: ModuleState;
}

export type MemoryAddressLookup = Record<string, number>;
