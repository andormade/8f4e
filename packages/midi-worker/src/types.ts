export interface MidiModuleAddresses {
	moduleId: string;
	noteWordAddress: number | undefined;
	channelWordAddress: number | undefined;
	noteOnOffWordAddress: number | undefined;
	velocityWordAddress: number | undefined;
}

export interface MidiCCModuleAddresses {
	moduleId: string;
	valueWordAddress: number | undefined;
	channelWordAddress: number | undefined;
	selectedCCWordAddress: number | undefined;
}

export interface RNBOModule {
	patcherId: string;
	paramAdresses: number[];
}
