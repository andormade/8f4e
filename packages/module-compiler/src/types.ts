export const enum ArgumentType {
	LITERAL = 'literal',
	IDENTIFIER = 'identifier',
}

export type Argument = { type: ArgumentType.LITERAL; value: number } | { type: ArgumentType.IDENTIFIER; value: string };

export type AST = Array<{ instruction: string; arguments: Array<Argument> }>;

export enum MemoryTypes {
	DYNAMIC_ARRAY,
	STATIC_ARRAY,
	ARRAY_SIZE,
	INPUT_POINTER,
	OUTPUT,
	PRIVATE,
	NUMBER,
	ARRAY,
}

export type MemoryMap = Array<{
	type: MemoryTypes;
	address: number;
	byteAddress: number;
	id: string;
	size: number;
	default: number | number[];
	usage: number;
}>;
