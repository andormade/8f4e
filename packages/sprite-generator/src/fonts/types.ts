export enum Glyph {
	SPACE,
	DOT,
	ARROW_TOP,
	ARROW_RIGHT,
	ARROW_BOTTOM,
	ARROW_LEFT,
	CONNECTOR_LEFT,
	CONNECTOR_RIGHT,
	SWITCH_KNOB,
}
