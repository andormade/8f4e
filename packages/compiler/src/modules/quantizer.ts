import {
	Instruction,
	Type,
	block,
	br_if,
	createFunctionBody,
	createLocalDeclaration,
	i32const,
	i32load,
	i32store,
	ifelse,
	localGet,
	localSet,
	loop,
} from '@8f4e/bytecode-utils';
import { ModuleGenerator, ModuleStateInserter, ModuleStateExtractor, MemoryTypes } from '../types';
import { I32_SIGNED_LARGEST_NUMBER } from '../consts';

export enum Memory {
	INPUT_POINTER,
	OUTPUT,
	ALLOCATED_NOTES,
	NUMBER_OF_NOTES,
	FIRST_NOTE,
}

enum Locals {
	BEST_MACTHING_VALUE,
	DIFFERENCE,
	INPUT,
	NOTE_MEMORY_POINTER,
	NOTE_VALUE,
	SMALLEST_DIFFERENCE,
	NOTES_END_ADDRESS_POINTER,
	__LENGTH,
}

const abs = registerIndex => [
	...localGet(registerIndex),
	...i32const(0),
	Instruction.I32_LT_S,
	...ifelse(Type.I32, [...i32const(0), ...localGet(registerIndex), Instruction.I32_SUB], [...localGet(registerIndex)]),
	...localSet(registerIndex),
];

interface QuantizerState {
	activeNotes: number[];
}

export const extractState: ModuleStateExtractor<QuantizerState> = function extractState(
	memoryBuffer,
	moduleAddress
): QuantizerState {
	const firstNoteAddress = moduleAddress + Memory.FIRST_NOTE;
	const numberOfNotesAddress = moduleAddress + Memory.NUMBER_OF_NOTES;
	return {
		activeNotes: Array.from(
			memoryBuffer.slice(firstNoteAddress, firstNoteAddress + memoryBuffer[numberOfNotesAddress])
		),
	};
};

export const insertState: ModuleStateInserter<QuantizerState> = function insertState(
	state,
	memoryBuffer,
	moduleAddress
): void {
	const firstNoteAddress = moduleAddress + Memory.FIRST_NOTE;
	const numberOfNotesAddress = moduleAddress + Memory.NUMBER_OF_NOTES;
	const allocatedNotesAddress = moduleAddress + Memory.ALLOCATED_NOTES;
	const allocatedNotes = memoryBuffer[allocatedNotesAddress];

	state.activeNotes.slice(0, allocatedNotes).forEach((note, index) => {
		memoryBuffer[firstNoteAddress + index] = note;
	});

	memoryBuffer[numberOfNotesAddress] = Math.min(state.activeNotes.length, allocatedNotes);
};

export interface Config {
	allocatedNotes?: number;
}

const quantizer: ModuleGenerator<Config> = function (moduleId, offset, config = {}) {
	const { allocatedNotes = 12 } = config;

	const functionBody = createFunctionBody(
		[createLocalDeclaration(Type.I32, Locals.__LENGTH)],
		[
			// Load the input value from the memory and put it into a register.
			...i32const(offset.byte(Memory.INPUT_POINTER)),
			...i32load(),
			...i32load(),
			...localSet(Locals.INPUT),

			// Calculate the address of the last note.
			...i32const(offset.byte(Memory.NUMBER_OF_NOTES)),
			...i32load(),
			...i32const(Int32Array.BYTES_PER_ELEMENT),
			Instruction.I32_MUL,
			...i32const(offset.byte(Memory.FIRST_NOTE)),
			Instruction.I32_ADD,
			...localSet(Locals.NOTES_END_ADDRESS_POINTER),

			// Set the smallest difference to the largest 32bit signed number.
			...i32const(I32_SIGNED_LARGEST_NUMBER),
			...localSet(Locals.SMALLEST_DIFFERENCE),

			// Set the note memory pointer to the start address
			...i32const(offset.byte(Memory.FIRST_NOTE)),
			...localSet(Locals.NOTE_MEMORY_POINTER),

			...block(
				Type.VOID,
				loop(Type.VOID, [
					// Break if the memory pointer would overflow.
					...localGet(Locals.NOTE_MEMORY_POINTER),
					...localGet(Locals.NOTES_END_ADDRESS_POINTER),
					Instruction.I32_GE_U,
					...br_if(1),

					// Load a note value from the memory.
					...localGet(Locals.NOTE_MEMORY_POINTER),
					...i32load(),
					...localSet(Locals.NOTE_VALUE),

					// Calculate difference between input and the note.
					...localGet(Locals.NOTE_VALUE),
					...localGet(Locals.INPUT),
					Instruction.I32_SUB,
					...localSet(Locals.DIFFERENCE),

					// Abs
					...abs(Locals.DIFFERENCE),

					// Compare with the smallest difference.
					...localGet(Locals.DIFFERENCE),
					...localGet(Locals.SMALLEST_DIFFERENCE),
					Instruction.I32_LE_S,
					...ifelse(Type.VOID, [
						// If it's actually smaller than the previusly recorded
						// smallest difference, then overwirte it and save the
						// current note value.
						...localGet(Locals.DIFFERENCE),
						...localSet(Locals.SMALLEST_DIFFERENCE),
						...localGet(Locals.NOTE_VALUE),
						...localSet(Locals.BEST_MACTHING_VALUE),
					]),

					// Increment the note memory pointer by 4 (i32 is 4 bytes)
					...localGet(Locals.NOTE_MEMORY_POINTER),
					...i32const(Int32Array.BYTES_PER_ELEMENT),
					Instruction.I32_ADD,
					...localSet(Locals.NOTE_MEMORY_POINTER),
				])
			),

			// Prepare memory address for storing the output value.
			...i32const(offset.byte(Memory.OUTPUT)),
			...localGet(Locals.BEST_MACTHING_VALUE),
			...i32store(),
		]
	);

	return {
		moduleId,
		functionBody,
		byteAddress: offset.byte(0),
		wordAddress: offset.word(0),
		memoryMap: [
			{ type: MemoryTypes.INPUT_POINTER, address: Memory.INPUT_POINTER, id: 'in', default: 0 },
			{ type: MemoryTypes.OUTPUT, address: Memory.OUTPUT, id: 'out', default: 0 },
			{
				type: MemoryTypes.PRIVATE,
				address: Memory.ALLOCATED_NOTES,
				id: 'allocatedNotes',
				default: allocatedNotes,
			},
			{ type: MemoryTypes.NUMBER, address: Memory.NUMBER_OF_NOTES, id: 'numberOfNotes', default: 0 },
			{
				type: MemoryTypes.DYNAMIC_ARRAY,
				sizePointer: Memory.NUMBER_OF_NOTES,
				maxSize: allocatedNotes,
				address: Memory.FIRST_NOTE,
				id: '',
				default: new Array(allocatedNotes).fill(-1),
			},
		],
	};
};

export default quantizer;
