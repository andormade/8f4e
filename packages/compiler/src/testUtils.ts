import wabt from 'wabt';

import {
	createCodeSection,
	createExportSection,
	createFunctionExport,
	createFunctionSection,
	createFunctionType,
	createImportSection,
	createMemoryImport,
	createTypeSection,
} from './wasmUtils/sectionHelpers';
import { FunctionBody } from './wasmUtils/typeHelpers';
import { CompiledModule, TestModule } from './types';
import { WORD_LENGTH } from './consts';

import { compileModules, getInitialMemory } from '.';

const HEADER = [0x00, 0x61, 0x73, 0x6d];
const VERSION = [0x01, 0x00, 0x00, 0x00];

export function createSingleFunctionWASMProgram(functionBody: FunctionBody): Uint8Array {
	return Uint8Array.from([
		...HEADER,
		...VERSION,
		...createTypeSection([createFunctionType([], [])]),
		...createImportSection([createMemoryImport('js', 'memory')]),
		...createFunctionSection([0x00]),
		...createExportSection([createFunctionExport('test', 0x00)]),
		...createCodeSection([functionBody]),
	]);
}

export function setInitialMemory(memory: DataView, module: CompiledModule): void {
	const initialMemory = getInitialMemory(module);
	for (let i = 0; i < initialMemory.length; i++) {
		if (Number.isInteger(initialMemory[i])) {
			memory.setInt32(i * 4, initialMemory[i], true);
		} else {
			memory.setFloat32(i * 4, initialMemory[i], true);
		}
	}
}

export async function createTestModule(sourceCode: string): Promise<TestModule> {
	let allocatedMemoryForTestData = 0;

	const module: CompiledModule = compileModules([{ code: sourceCode.split('\n') }])[0];
	const program = createSingleFunctionWASMProgram(module.functionBody);
	const memoryRef = new WebAssembly.Memory({ initial: 1 });
	const dataView = new DataView(memoryRef.buffer);
	const memoryBuffer = new Int32Array(memoryRef.buffer);
	let instance: WebAssembly.Instance | undefined;
	let wat = '';

	try {
		const webAssemblyInstantiatedSource = await WebAssembly.instantiate(program, {
			js: {
				memory: memoryRef,
			},
		});
		instance = webAssemblyInstantiatedSource.instance;

		wat = await new Promise(resolve => {
			wabt().then(_wabt => {
				const module = _wabt.readWasm(program, {});
				resolve(module.toText({}));
			});
		});
	} catch (error) {
		console.log(error);
	}

	const reset = () => {
		setInitialMemory(dataView, module);

		// Clear the test data that was out of the module's scope.
		for (let i = module.memoryWordSize; i < module.memoryWordSize + allocatedMemoryForTestData; i++) {
			memoryBuffer[i] = 0;
		}
		allocatedMemoryForTestData = 0;
	};

	reset();

	const test = (instance?.exports.test ||
		function () {
			return;
		}) as CallableFunction;

	const memoryGet = (address: string | number): number | undefined => {
		if (typeof address === 'string') {
			const memoryItem = module.memoryMap.get(address);

			if (!memoryItem) {
				return;
			}

			if (memoryItem.isInteger) {
				return dataView.getInt32(memoryItem.byteAddress, true);
			} else {
				return dataView.getFloat32(memoryItem.byteAddress, true);
			}
		}

		return memoryBuffer[address];
	};

	const memorySet = (address: string | number, value: number | number[]): void => {
		if (typeof address === 'string') {
			const memoryItem = module.memoryMap.get(address);
			if (!memoryItem) {
				return;
			}

			if (typeof value === 'number') {
				if (Number.isInteger(value)) {
					dataView.setInt32(memoryItem.byteAddress, value, true);
				} else {
					dataView.setFloat32(memoryItem.byteAddress, value, true);
				}
			} else {
				for (let i = 0; i < value.length; i++) {
					if (Number.isInteger(value[i])) {
						dataView.setInt32(memoryItem.byteAddress + i * 4, value[i], true);
					} else {
						dataView.setFloat32(memoryItem.byteAddress + i * 4, value[i], true);
					}
				}
			}
			return;
		}

		if (typeof value === 'number') {
			if (Number.isInteger(value)) {
				dataView.setInt32(address * 4, value, true);
			} else {
				dataView.setFloat32(address * 4, value, true);
			}
		} else {
			for (let i = 0; i < value.length; i++) {
				if (Number.isInteger(value[i])) {
					dataView.setInt32(address * 4 + i * 4, value[i], true);
				} else {
					dataView.setFloat32(address * 4 + i * 4, value[i], true);
				}
			}
		}
	};

	const getByteAddress = (address: string | number): number | undefined => {
		if (typeof address === 'string') {
			const memoryItem = module.memoryMap.get(address);
			if (!memoryItem) {
				return;
			}

			return memoryItem.byteAddress;
		}

		return address * WORD_LENGTH;
	};

	const allocMemoryForPointer = (address: string | number): number => {
		const firstFreeAddress = module.memoryWordSize + allocatedMemoryForTestData;
		memorySet(address, firstFreeAddress * WORD_LENGTH);
		allocatedMemoryForTestData++;
		return firstFreeAddress;
	};

	// @ts-ignore
	memoryBuffer.get = memoryGet;
	// @ts-ignore
	memoryBuffer.set = memorySet;
	// @ts-ignore
	memoryBuffer.allocMemoryForPointer = allocMemoryForPointer;
	// @ts-ignore
	memoryBuffer.byteAddress = getByteAddress;

	return {
		// @ts-ignore
		memory: memoryBuffer,
		test,
		reset,
		wat,
		program,
		memoryMap: module.memoryMap,
		ast: module.ast,
	};
}
