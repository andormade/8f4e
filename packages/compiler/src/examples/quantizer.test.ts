import quantizer from './quantizer.asm';

import { createTestModule } from '../testUtils';
import { TestModule } from '../types';

let testModule: TestModule;

const fixtures: [input: number, notes: number[], output: number][] = [
	[1000, [], 0],
	[800, [], 0],
	[400, [], 0],

	[0, [900], 900],
	[1000, [900], 900],
	[800, [900], 900],
	[400, [900], 900],

	[900, [1000, 500], 1000],
	[800, [1000, 500], 1000],
	[400, [1000, 500, 100], 500],
	[500, [1000, 500, 100], 500],
	[600, [1000, 500, 100], 500],

	[-900, [-1000, 500], -1000],
	[-800, [-1000, 500], -1000],
	[-400, [-1000, 500], -1000],
];

describe('functional tests', () => {
	beforeAll(async () => {
		testModule = await createTestModule(quantizer());
	});

	beforeEach(() => {
		testModule.reset();
	});

	test('if the generated ast matches with the snapshot', () => {
		expect(testModule.ast).toMatchSnapshot();
	});

	test('if the wat code matches with the snapshot', () => {
		expect(testModule.wat).toMatchSnapshot();
	});

	test('if the generated memory map matches with the snapshot', () => {
		expect(testModule.memoryMap).toMatchSnapshot();
	});

	test('quantizer module', () => {
		const { memory, test } = testModule;
		test();
		expect(memory[1]).toBe(0);
	});

	test.each(fixtures)('given %p as input, %p as active notes, the expected output is %p', (input, notes, output) => {
		const { memory, test } = testModule;

		const inAddress = memory.allocMemoryForPointer('in');

		memory.set('numberOfNotes', notes.length);

		memory.set('notes', notes);

		memory.set(inAddress, input);
		test();
		expect(memory.get('out')).toBe(output);
	});
});
