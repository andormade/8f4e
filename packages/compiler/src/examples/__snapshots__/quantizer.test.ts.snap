// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`functional tests if the generated ast matches with the snapshot 1`] = `
Array [
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "quantizer",
      },
    ],
    "instruction": "module",
    "lineNumber": 0,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "in",
      },
      Object {
        "isInteger": true,
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "int*",
    "lineNumber": 2,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "out",
      },
      Object {
        "isInteger": true,
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "int",
    "lineNumber": 3,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "allocatedNotes",
      },
      Object {
        "isInteger": true,
        "type": "literal",
        "value": 12,
      },
    ],
    "instruction": "int",
    "lineNumber": 4,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "numberOfNotes",
      },
      Object {
        "isInteger": true,
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "int",
    "lineNumber": 5,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "notes",
      },
      Object {
        "isInteger": true,
        "type": "literal",
        "value": 12,
      },
      Object {
        "isInteger": true,
        "type": "literal",
        "value": -1,
      },
    ],
    "instruction": "int[]",
    "lineNumber": 6,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "int",
      },
      Object {
        "type": "identifier",
        "value": "bestMatchingValue",
      },
    ],
    "instruction": "local",
    "lineNumber": 8,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "int",
      },
      Object {
        "type": "identifier",
        "value": "difference",
      },
    ],
    "instruction": "local",
    "lineNumber": 9,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "int",
      },
      Object {
        "type": "identifier",
        "value": "_in",
      },
    ],
    "instruction": "local",
    "lineNumber": 10,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "int",
      },
      Object {
        "type": "identifier",
        "value": "noteMemoryPointer",
      },
    ],
    "instruction": "local",
    "lineNumber": 11,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "int",
      },
      Object {
        "type": "identifier",
        "value": "noteValue",
      },
    ],
    "instruction": "local",
    "lineNumber": 12,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "int",
      },
      Object {
        "type": "identifier",
        "value": "smallestDifference",
      },
    ],
    "instruction": "local",
    "lineNumber": 13,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "int",
      },
      Object {
        "type": "identifier",
        "value": "notesEndAddressPointer",
      },
    ],
    "instruction": "local",
    "lineNumber": 14,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "*in",
      },
    ],
    "instruction": "push",
    "lineNumber": 16,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_in",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 17,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "numberOfNotes",
      },
    ],
    "instruction": "push",
    "lineNumber": 20,
  },
  Object {
    "arguments": Array [
      Object {
        "isInteger": true,
        "type": "literal",
        "value": 4,
      },
    ],
    "instruction": "push",
    "lineNumber": 21,
  },
  Object {
    "arguments": Array [],
    "instruction": "mul",
    "lineNumber": 22,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "&notes",
      },
    ],
    "instruction": "push",
    "lineNumber": 23,
  },
  Object {
    "arguments": Array [],
    "instruction": "add",
    "lineNumber": 24,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "notesEndAddressPointer",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 25,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "I32_SIGNED_LARGEST_NUMBER",
      },
    ],
    "instruction": "push",
    "lineNumber": 28,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "smallestDifference",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 29,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "&notes",
      },
    ],
    "instruction": "push",
    "lineNumber": 32,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "noteMemoryPointer",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 33,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "void",
      },
    ],
    "instruction": "block",
    "lineNumber": 35,
  },
  Object {
    "arguments": Array [],
    "instruction": "loop",
    "lineNumber": 36,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "noteMemoryPointer",
      },
    ],
    "instruction": "push",
    "lineNumber": 38,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "notesEndAddressPointer",
      },
    ],
    "instruction": "push",
    "lineNumber": 39,
  },
  Object {
    "arguments": Array [],
    "instruction": "greaterOrEqualUnsigned",
    "lineNumber": 40,
  },
  Object {
    "arguments": Array [
      Object {
        "isInteger": true,
        "type": "literal",
        "value": 1,
      },
    ],
    "instruction": "branchIfTrue",
    "lineNumber": 41,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "noteMemoryPointer",
      },
    ],
    "instruction": "push",
    "lineNumber": 44,
  },
  Object {
    "arguments": Array [],
    "instruction": "load",
    "lineNumber": 45,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "noteValue",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 46,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "noteValue",
      },
    ],
    "instruction": "push",
    "lineNumber": 49,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_in",
      },
    ],
    "instruction": "push",
    "lineNumber": 50,
  },
  Object {
    "arguments": Array [],
    "instruction": "sub",
    "lineNumber": 51,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "difference",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 52,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "difference",
      },
    ],
    "instruction": "push",
    "lineNumber": 55,
  },
  Object {
    "arguments": Array [
      Object {
        "isInteger": true,
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "push",
    "lineNumber": 56,
  },
  Object {
    "arguments": Array [],
    "instruction": "lessThan",
    "lineNumber": 57,
  },
  Object {
    "arguments": Array [],
    "instruction": "if",
    "lineNumber": 58,
  },
  Object {
    "arguments": Array [
      Object {
        "isInteger": true,
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "push",
    "lineNumber": 59,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "difference",
      },
    ],
    "instruction": "push",
    "lineNumber": 60,
  },
  Object {
    "arguments": Array [],
    "instruction": "sub",
    "lineNumber": 61,
  },
  Object {
    "arguments": Array [],
    "instruction": "else",
    "lineNumber": 62,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "difference",
      },
    ],
    "instruction": "push",
    "lineNumber": 63,
  },
  Object {
    "arguments": Array [],
    "instruction": "end",
    "lineNumber": 64,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "difference",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 65,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "difference",
      },
    ],
    "instruction": "push",
    "lineNumber": 68,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "smallestDifference",
      },
    ],
    "instruction": "push",
    "lineNumber": 69,
  },
  Object {
    "arguments": Array [],
    "instruction": "lessOrEqual",
    "lineNumber": 70,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "void",
      },
    ],
    "instruction": "if",
    "lineNumber": 71,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "difference",
      },
    ],
    "instruction": "push",
    "lineNumber": 75,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "smallestDifference",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 76,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "noteValue",
      },
    ],
    "instruction": "push",
    "lineNumber": 77,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "bestMatchingValue",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 78,
  },
  Object {
    "arguments": Array [],
    "instruction": "end",
    "lineNumber": 79,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "noteMemoryPointer",
      },
    ],
    "instruction": "push",
    "lineNumber": 81,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "WORD_SIZE",
      },
    ],
    "instruction": "push",
    "lineNumber": 82,
  },
  Object {
    "arguments": Array [],
    "instruction": "add",
    "lineNumber": 83,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "noteMemoryPointer",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 84,
  },
  Object {
    "arguments": Array [],
    "instruction": "loopEnd",
    "lineNumber": 85,
  },
  Object {
    "arguments": Array [],
    "instruction": "end",
    "lineNumber": 86,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "&out",
      },
    ],
    "instruction": "push",
    "lineNumber": 89,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "bestMatchingValue",
      },
    ],
    "instruction": "push",
    "lineNumber": 90,
  },
  Object {
    "arguments": Array [],
    "instruction": "store",
    "lineNumber": 91,
  },
  Object {
    "arguments": Array [],
    "instruction": "moduleEnd",
    "lineNumber": 93,
  },
]
`;

exports[`functional tests if the generated memory map matches with the snapshot 1`] = `
Map {
  "in" => Object {
    "byteAddress": 0,
    "default": 0,
    "id": "in",
    "isInteger": true,
    "isPointer": true,
    "isPointingToInteger": true,
    "isPointingToPointer": false,
    "lineNumber": 2,
    "type": "int*",
    "wordAddress": 0,
    "wordSize": 1,
  },
  "out" => Object {
    "byteAddress": 4,
    "default": 0,
    "id": "out",
    "isInteger": true,
    "isPointer": false,
    "isPointingToInteger": false,
    "isPointingToPointer": false,
    "lineNumber": 3,
    "type": "int",
    "wordAddress": 1,
    "wordSize": 1,
  },
  "allocatedNotes" => Object {
    "byteAddress": 8,
    "default": 12,
    "id": "allocatedNotes",
    "isInteger": true,
    "isPointer": false,
    "isPointingToInteger": false,
    "isPointingToPointer": false,
    "lineNumber": 4,
    "type": "int",
    "wordAddress": 2,
    "wordSize": 1,
  },
  "numberOfNotes" => Object {
    "byteAddress": 12,
    "default": 0,
    "id": "numberOfNotes",
    "isInteger": true,
    "isPointer": false,
    "isPointingToInteger": false,
    "isPointingToPointer": false,
    "lineNumber": 5,
    "type": "int",
    "wordAddress": 3,
    "wordSize": 1,
  },
  "notes" => Object {
    "byteAddress": 16,
    "default": Array [
      -1,
      -1,
      -1,
      -1,
      -1,
      -1,
      -1,
      -1,
      -1,
      -1,
      -1,
      -1,
    ],
    "id": "notes",
    "isInteger": true,
    "isPointer": false,
    "isPointingToInteger": false,
    "isPointingToPointer": false,
    "lineNumber": 6,
    "type": "int",
    "wordAddress": 4,
    "wordSize": 12,
  },
  "__loopErrorSignaler" => Object {
    "byteAddress": 64,
    "default": -1,
    "id": "__loopErrorSignaler",
    "isInteger": true,
    "isPointer": false,
    "isPointingToInteger": false,
    "isPointingToPointer": false,
    "lineNumber": 1,
    "type": "int",
    "wordAddress": 16,
    "wordSize": 1,
  },
}
`;

exports[`functional tests if the wat code matches with the snapshot 1`] = `
"(module
  (type (;0;) (func))
  (import \\"js\\" \\"memory\\" (memory (;0;) 1))
  (func (;0;) (type 0)
    (local i32 i32 i32 i32 i32 i32 i32 i32)
    i32.const 0
    i32.load
    i32.load
    local.set 2
    i32.const 12
    i32.load
    i32.const 4
    i32.mul
    i32.const 16
    i32.add
    local.set 6
    i32.const 2147483647
    local.set 5
    i32.const 16
    local.set 3
    block  ;; label = @1
      i32.const 0
      local.set 7
      block  ;; label = @2
        loop  ;; label = @3
          local.get 7
          i32.const 1000
          i32.ge_s
          if  ;; label = @4
            i32.const 64
            i32.const 36
            i32.store
            br 2 (;@2;)
          end
          local.get 7
          i32.const 1
          i32.add
          local.set 7
          local.get 3
          local.get 6
          i32.ge_u
          br_if 1 (;@2;)
          local.get 3
          i32.load
          local.set 4
          local.get 4
          local.get 2
          i32.sub
          local.set 1
          local.get 1
          i32.const 0
          i32.lt_s
          if (result i32)  ;; label = @4
            i32.const 0
            local.get 1
            i32.sub
          else
            local.get 1
          end
          local.set 1
          local.get 1
          local.get 5
          i32.le_s
          if  ;; label = @4
            local.get 1
            local.set 5
            local.get 4
            local.set 0
          end
          local.get 3
          i32.const 4
          i32.add
          local.set 3
          br 0 (;@3;)
        end
      end
    end
    i32.const 4
    local.get 0
    i32.store)
  (export \\"test\\" (func 0)))
"
`;
